### DISCUSSION TITLE
Ray cast not finding hit with static triangle mesh

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#1000] [Q&A] [answered]
Been trying to implement a raycasting function using the examples in places like
https://github.com/jrouwe/JoltPhysics/blob/e8a99bfa0c60997eb4a7c22bd5c34f36a87dc632/UnitTests/Physics/RayShapeTests.cpp#L325

But the hit isn't registering... the collection size returns a zero. Even if I want to query the entire physics system, do I need to supply filters to the CastRay() function, or is there something else I'm not understanding about how this works?

Here is an example, of trying to find walls with the camera, where I'm getting one of these zero collection sizes (camera 'ANGLE' vector is the look vector of the camera in this scene). There is a wall where you see the floor cut off here:
![image](https://github.com/jrouwe/JoltPhysics/assets/159929665/62c78322-acc3-4e2c-be53-7baf1549adb3)

`

bool CollisionHandler::GetClosestFromRay(int colFlags, const Vector::vector_t *startPoint, const Vector::vector_t *endPoint, Vector::vector_t *colPoint, Vector::vector_t *colNormal, WObj::WObj *excludeObj, const float dotLimit)
{
	JPH::Vec3 start(startPoint->x, startPoint->y, startPoint->z);
	JPH::Vec3 end(endPoint->x, endPoint->y, endPoint->z);

	Vector::vector_t direction;
	Vector::Sub(endPoint, startPoint, &direction);
	Vector::Normalize(&direction);
	JPH::Vec3 dir(direction.x, direction.y, direction.z);

	// Create ray
	JPH::RRayCast ray { start, dir };

	JPH::RayCastSettings settings;
	settings.mBackFaceMode = JPH::EBackFaceMode::CollideWithBackFaces;

	// Cast ray
	JPH::AllHitCollisionCollector<JPH::CastRayCollector> collector;
	mPhysicsSystem->GetNarrowPhaseQuery().CastRay(ray, settings, collector);
	collector.Sort();

	bool hasHit = collector.HadHit();
	if (hasHit)
	{
		int closestIndex = -1;
		float closestDist = -1.0f;

		JPH::BodyInterface &bi = mPhysicsSystem->GetBodyInterface();
		for (int i = 0; i < collector.mHits.size(); i++)
		{
			const JPH::RayCastResult &hit = collector.mHits[i];
			.......

`

**Comment by jrouwe** [comment, marked as answer]
By doing:

```
Vector::Normalize(&direction);
```

You effectively make your ray 1 unit long, so that's probably why you're not hitting anything.

**Reply by SSNTails** [reply]
Oh, so a ray is really a line segment with an endpoint, not a ray (point with normalized direction)?

It may be helpful to modify the samples to make that a bit clearer. That's where I was getting tripped up, as the variable names in the samples don't really make it clear that it's a non-normalized distance value:

https://github.com/jrouwe/JoltPhysics/blob/e8a99bfa0c60997eb4a7c22bd5c34f36a87dc632/Samples/SamplesApp.cpp#L1037
https://github.com/jrouwe/JoltPhysics/blob/e8a99bfa0c60997eb4a7c22bd5c34f36a87dc632/UnitTests/Physics/TransformedShapeTests.cpp#L85
